src/Build/Graph/GraphBuilder.cs

```diff
@@ -621,9 +621,42 @@ internal class GraphEdges
                     ErrorUtilities.VerifyThrow(ReferenceItems.TryGetValue(key, out ProjectItemInstance referenceItem), "All requested keys should exist");
                     return referenceItem;
                 }
+            }
+
+            public void AddOrUpdateEdge((ProjectGraphNode node, ProjectGraphNode reference) key, ProjectItemInstance edge)
+            {
+                ReferenceItems.AddOrUpdate(
+                    key,
+                    addValueFactory: static ((ProjectGraphNode node, ProjectGraphNode reference) key, ProjectItemInstance referenceItem) => referenceItem,
+                    updateValueFactory: static ((ProjectGraphNode node, ProjectGraphNode reference) key, ProjectItemInstance existingItem, ProjectItemInstance newItem) =>
+                    {
+                        string existingTargetsMetadata = existingItem.GetMetadataValue(ItemMetadataNames.ProjectReferenceTargetsMetadataName);
+                        string newTargetsMetadata = newItem.GetMetadataValue(ItemMetadataNames.ProjectReferenceTargetsMetadataName);
+
+                        // Bail out of the targets are the same.
+                        if (existingTargetsMetadata.Equals(newTargetsMetadata, StringComparison.OrdinalIgnoreCase))
```

Comment:
Looks like MSBuild isn't case sensitive for Target names. Here's what it looks like for a non-graph build:

<img width="562" alt="image" src="https://user-images.githubusercontent.com/6445614/179619983-b41cd3ab-ae40-4b46-b845-43860160716b.png">

So I think using `StringComparison.OrdinalIgnoreCase` is fine here.

Also duplicate targets in the list should be handled just fine anyway; this was more of an optimization to avoid allocating when merging things which clearly are equivalent.